packages/store/src/StoreCore.sol

```diff
@@ -554,12 +542,32 @@ library StoreCoreInternal {
   ) internal {
     uint8 dynamicSchemaIndex = schemaIndex - uint8(valueSchema.numStaticFields());
 
-    // Update the dynamic data length
-    _setDynamicDataLengthAtIndex(tableId, key, dynamicSchemaIndex, data.length);
+    // Load dynamic data length from storage
+    uint256 dynamicSchemaLengthSlot = _getDynamicDataLengthLocation(tableId, key);
+    PackedCounter encodedLengths = PackedCounter.wrap(Storage.load({ storagePointer: dynamicSchemaLengthSlot }));
+
+    // Update the encoded length
+    uint256 oldFieldLength = encodedLengths.atIndex(dynamicSchemaIndex);
+    encodedLengths = encodedLengths.setAtIndex(dynamicSchemaIndex, data.length);
+    // Set the new lengths
+    Storage.store({ storagePointer: dynamicSchemaLengthSlot, data: encodedLengths.unwrap() });
 
     // Store the provided value in storage
     uint256 dynamicDataLocation = _getDynamicDataLocation(tableId, key, dynamicSchemaIndex);
     Storage.store({ storagePointer: dynamicDataLocation, offset: 0, data: data });
+
+    // Prepare data for the splice event
+    uint256 start;
+    unchecked {
+      // (safe because it's a few uint40 values, which can't overflow uint48)
+      start = valueSchema.staticDataLength() + 32;
+      for (uint8 i; i < dynamicSchemaIndex; i++) {
+        start += encodedLengths.atIndex(i);
+      }
+    }
+    uint256 deleteCount = oldFieldLength;
+    // Emit event to notify indexers
+    emit StoreCore.StoreSpliceRecord(tableId, key, uint48(start), uint40(deleteCount), data);
```

Comment:
Is it worth generalizing this in an internal util function, since very similar logic is used for `_setDynamicField`, `_pushToDynamicField`, `_popFromDynamicField` and `_setDynamicFieldItem`? I feel like might declutter the code a bit, make it easier to audit and easier to write tests for this logic in isolation

Reply:
I've been thinking about this for a while now. I think the only efficient way to declutter is generalize it in a renderer function, and render StoreCore. Using solidity functions means more gas because the optimizer won't inline any of it and we'll need to return many vars

Reply:
Or we could not worry about gas I guess, it's probably less than 1000 (haven't tested specifics). Though a bunch of return vars may still be ugly compared to a renderer with conditionals in the relevant places

Reply:
@alvrs do you want declutter in this PR? it feels like it'd be better in a separate one for easier reverts and gas comparisons

Reply:
can try to refactor in a separate PR